torch/csrc/jit/codegen/cuda/runtime/helpers.cu

```diff
@@ -307,3 +307,25 @@ float pow(float a, int64_t b) {
 double pow(double a, int64_t b) {
   return pow(a, (double)b);
 }
+
+template <int size, int align = size>
+struct alignas(align) TypelessData {
+  int8_t data[size];
+
+  template <typename T, std::enable_if_t<sizeof(T) == size, int> _ = 0>
+  TypelessData(T x) {
+    memcpy(data, &x, size);
```

Comment:
@csarofeen recently mentioned he saw some strange results when `memcpy` was used with register values. Just a for-loop copy would be fine too.

Reply:
I am not sure about this. If it was a copy of two arrays of the same type, then a for loop could be pretty good, but for the case here, in order to do for loop copy, I will have to use `reinterpret_cast` to treat `&x` as a different type, which is not a recommended way to do type punning according to https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#how-do-we-type-pun-correctly
Indeed @ptrblck fixed a bug caused by this strict-aliasing issue https://github.com/pytorch/pytorch/pull/66194 caused by `reinterpret_cast` on ARM. So I have a feeling that `memcpy` should be a safer solution than `reinterpret_cast`.

Reply:
+1 to @zasdfgbnm's explanation.

> he saw some strange results when memcpy was used with register values

That's also true, and we should investigate it separately. I can check @csarofeen's code again (didn't see any issues so far).

Reply:
Oh, I see. I wonder if it would also mean our vectorized loads and stores with, e.g., `reinterpret_cast<float4*>` were considered unsafe.

https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/runtime/array.cu#L27-L60

Interestingly, he saw an error with `memcpy` and a workaround was to use `reinterpret_cast`.

Reply:
> Interestingly, he saw an error with `memcpy` and a workaround was to use `reinterpret_cast`.

Exactly! And so far I don't understand why that's the case. Thanks for sending the code snippet btw.

Reply:
How vectorized loads and stores are done in the PyTorch eager mode? Doesn't it use `reinterpret_cast<vector_type*>`?

Reply:
> How vectorized loads and stores are done in the PyTorch eager mode? Doesn't it use `reinterpret_cast<vector_type*>`?

Unfortunately it does https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/MemoryAccess.cuh#L226 :(

Reply:
The problem only happens when a tensor is loaded and stored in an inconsistent way, i.e., a vectorized store followed by a non-vectorized load, or a non-vectorized store followed by a vectorized load. And it only matters when it happens in a single CUDA kernel. I wonder when such patterns would appear. Unlikely?

Reply:
When memcpy was used for our vectorized loads I think nvrtc had a hard time understanding the size in registers that was required for some reason. I think it was effectively not able to understand the side effects from the registers being referenced from the ptr + size. Reinterpret cast seemed to effectively indicate to the compiler "this is the set of registers I'm expecting". However, this is more of a hunch than a validated cause.

Reply:
Basically my opinion here is whatever works. If you want to see where it broke us you can revert the commenting of the memcpy in the runtime files for vectorized load/store and run the c++ tests.

Reply:
Marking this as resolved. Per offline discussion, we should prefer `reinterpret_cast`. And I also updated my PR using `reinterpret_cast`.